### ISSUE TITLE
Question: Events w/o reducers

### ISSUE THREAD

**Issue by brandonmp** [rangle/redux-segment#16]
Thanks for writing this--it's exactly what I was looking for.

I'm pretty new @ this so forgive if this question is misguided, but:

is there a simple way for redux-segment to track events that don't affect state, or that affect something in component-level local state? eg expanding an accordion, activating a dropdown, etc.  

my first thought is to just use the regular segment SDK functions within the components, but obviously that gets away from centralizing things.

**Comment by bertrandk** [comment]
Hey @brandonmp,

> ...is there a simple way for redux-segment to track events that don't affect state...

**Yes, no reducers needed. Skip to paragraph 2 below for more details.**

> ...or that affect something in component-level local state?

**Not without dispatching a token action, at the very least**

It's an interesting question. In general, the state _we care_ about is kept in the redux store itself. As a result, I don't think I've ever been in the position of needing to measure state changes stored at a component-level (to say nothing of the fact that we generally avoid stateful components). **And so, there's really no simple way to track events that affect component-level local state if no actual action is dispatched.**

That being said, actions don't actually need to affect the global (redux) state. Redux-segment is a middleware so it's totally oblivious to what your reducers are doing. You're completely free to dispatch an action that _no_ reducer handles. For instance, when expanding an accordion, you could dispatch an action before calling `setState` locally. That action could just be ignored by all reducers. In essence, you'd then be tracking an event that doesn't affect state. Whether this approach is better than just calling the regular SDK is up to you. I'd be inclined to argue that at least having this "token" action would make things more explicit (i.e. your tracking activities remain declarative, centralized and decoupled from the actual components triggering them).

Hope this helps. Let me now if I need to clarify anything.

**Comment by brandonmp** [comment]
wow, thanks for the thoughtful response @bertrandk -- github is generally pretty helpful, but often quite terse, so yours is refreshing.

> That being said, actions don't actually need to affect the global (redux) state

^^this idea reconciles a lot of redux 'best practices' i was having trouble jiving.  i couldn't quite reason where the line b/t redux-worthy state and local state was, but if i can perform side fx w/in actions (eg, segment tracking) w/o poking around in redux state, then it becomes a lot less important to get "local vs. app-level" right, since i can still house all my logic w/in actions.

> In general, the state _we care_ about is kept in the redux store itself. 

Can i ask you to clarify that a little? what're your metrics for caring about a particular piece of state enough to store it in redux?  

understand that this might not be germane to redux-segment, but curious if i'm reasoning about state correctly (or at least, in a fashion similar to people who know what they're doing :) )

**Comment by bertrandk** [comment]
> > In general, the state we care about is kept in the redux store itself.

> Can i ask you to clarify that a little? what're your metrics for caring about a particular piece of state enough to store it in redux?

That was rather more a term of art. In general, I care about all state. It's the very reason I'm using redux to begin with. State, especially the kind that changes over time, can be complex and the conditions under which it changes typically become harder to reason about as your application grows and evolves. Redux – specifically its restrictions – allows us to manage that complexity by making state changes predictable (even when dealing with asynchronous events). In order to realize as much of that promise as possible, it stands to reason that as much state as possible must live in the redux store. Typically then, you want your entire application state in the store. It not only gives you the obvious benefits (serializability, introspection, etc..) but also ensure the conditions under which that state changes adhere to the contracts necessary to make state management sane (immutability, pure function, etc...).

Now, there are times where you may opt to _not_ have some application state in the store. For instance, I may keep the default value of an input locally. If I were to somehow lose that state or if it were to get out of sync, I may not care so much. In other words, I don't care about state when it's not likely to lead to any non-trivial bugs. So in a sense, that's what I meant above. Other examples of this may include intermediate values for complex animations, localized performance critical code (physics engines or where in-place mutations make an order of magnitude difference in execution speed), etc... In all such scenarios, I'm making an explicit trade-off between simplicity/predictability and some other benefit. The complexity in those cases simply may be unavoidable at the time.

In your case, you'll have to make that judgement. I can easily imagine a lot of applications where a bug in the toggling of an accordion would lead to (severe) loss of function for users. The same can be said for dropdown state. Forms are especially tough, them being just about the most stateful things this side of databases, so it might be helpful to look into leveraging any of the multitude of redux form helpers. 

Anyways, sorry for the latency and hope the above is clear enough. I'm a little sleepy. 😪

**TLDR**: I store _all_ my state in the redux store until I have a compelling reason not to (e.g. good benchmarks, other system constraints). When I don't, it's explicit and done with care. After all, bad state is the source of a lot (most?) defects in a lot of software.

**Comment by lucasfeliciano** [comment]
Good issue.

And good answers
